agent/xds/clusters.go

```diff
@@ -561,7 +572,24 @@ func (s *Server) makeGatewayClusterWithConnectTimeout(clusterName string, cfgSna
 		},
 		// Having an empty config enables outlier detection with default config.
 		OutlierDetection: &envoycluster.OutlierDetection{},
-	}, nil
+	}
+
+	// Terminating gateways support TLS origination from files/SNI specified in their config entry
+	if cfgSnap.Kind == structs.ServiceKindTerminatingGateway {
+		// This should never happen but need to make sure we don't panic on "*service" below
+		if service == nil {
+			return nil, fmt.Errorf("failed to inject TLS context due to nil service")
```

Comment:
One option that is appealing to me would be to put this new logic a new function which sets `TLSContext` on a n `envoy.Cluster`. 

It would mean 2-3 call sites would have the second two function calls after each other (I believe the 3 could be made into 2 calls easily enough).

On the plus side it means that 3/6 calls don't need to pass nil for the extra parameter, and I think maybe this extra checks wont be necessary because it will only be called from the terminating gateway flow.

Reply:
Added this as well to avoid passing nil as an extra parameter.